lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1349,12 +1415,48 @@ class MongoClientInterface {
      * @param {Function} cb callback
      * @return {undefined}
      */
-    internalListObject(bucketName, params, extension, log, cb) {
+    internalListObjectV1(bucketName, params, extension, log, cb) {
         const c = this.getCollection(bucketName);
+        const getLatestVersion = this.getLatestVersion;
         let stream;
         if (!params.secondaryStreamParams) {
-            stream = new MongoReadStream(c, params.mainStreamParams, params.mongifiedSearch);
+            /**
+             * When listing masters only we can't just skip PHD
+             * we have to replace them with the latest version of
+             * the object.
+             * Here we use a trasform stream that we pipe with the
+             * mongo read steam and that checks and replaces the key
+             * read if it's a PHD
+             *  */
+            const resolvePhdKey = new Transform({
+                objectMode: true,
+                transform(obj, encoding, callback) {
+                    if (Version.isPHD(obj.value)) {
+                        const key = obj.key.slice(DB_PREFIXES.Master.length);
+                        getLatestVersion(c, key, BUCKET_VERSIONS.v1, log, (err, version) => {
+                            if (err) {
+                                log.error(
+                                    'internalListObjectV1: error while getting latest version of PHD key');
+                                callback(errors.InternalError);
```

Comment:
```suggestion
                                return callback(errors.InternalError);
```

I am not sure we want to run the function in case of an error, and return the latestVersion with an undefined value.

Reply:
IMO i think it's better to return an error instead of skipping the entry, we don't want to return a false listing result, i.e with missing objects

Reply:
yes but without the return, you will run the following code in case of error:

```js
                            MongoUtils.unserialize(version);
                            // we keep the master key and only replace the value
                            const latestVersion = {
                                key: obj.key,
                                value: JSON.stringify(version),
                            };
                            callback(null, latestVersion);
```

Right? In this case, you will call the callback twice.

Reply:
oh i guess i just misunderstood what you said, but yeah i agree with adding the return